scriptworker/artifacts.py

```diff
@@ -306,3 +306,39 @@ def get_expiration_arrow(context):
 
     await raise_future_exceptions(tasks)
     return files
+
+
+def get_upstream_artifacts(context):
+    """List the downloaded upstream artifacts.
+
+    Args:
+        context (scriptworker.context.Context): the scriptworker context.
+
+    Returns:
+        dict: the paths of uploaded artifacts, sorted per task_id
+
+    Raises:
+        scriptworker.exceptions.ScriptWorkerTaskException: when an artifact doesn't exist.
+    """
+    task_ids_and_relative_paths = [
+        (artifact_definition['taskId'], artifact_definition['paths'])
+        for artifact_definition in context.task['payload']['upstreamArtifacts']
+    ]
+
+    return {
+        task_id: [
+            _get_single_upstream_artifact(context, task_id, path)
+            for path in paths
+        ]
+        for task_id, paths in task_ids_and_relative_paths
+    }
+
+
+def _get_single_upstream_artifact(context, task_id, path):
```

Comment:
`link.cot_dir` could even be a `functools.partial` that pre-fills-out `context` and `task_id` for `get_single_upstream_artifact_full_path`.  `link.cot_path('task.json')` would then return `$work_dir/cot/$task_id/task.json`.  (This isn't required; just something I thought of.)

I did notice we use `link.cot_dir` to define the `parent_dir` to download the artifacts into: https://github.com/mozilla-releng/scriptworker/blob/master/scriptworker/cot/verify.py#L535  Maybe we allow for `path` to be `None` in `get_single_upstream_artifact_full_path` ?  (Or `get_cot_full_path` or something if that gets unwieldy?)